### PR TITLE
fix: re-fetch appData on talk hash invalidate

### PR DESCRIPTION
**Pull request by ShGKme**
### ☑️ Resolves

* Issue #198 

### 🖼️ Screenshots

No visual changes

### 🚧 Tasks

`AppData` includes `capabilities` and `userMetadata`.

Before `appData` was collected on the first login, then persist in `localStorage` until logout. It makes it invalid on Talk update or Talk settings update.

This PR:
- Adds talk hash and dirty flag to `appData`
- Adds `appData.service` with functions to re-fetch capabilities and userMetadata
- Adds more utils methods in `AppData`
- Restores `appData` during the **Welcome** screen, without dirty `executeJavascript` in the main process
  - And re-fetch if talk hash is dirty
- Initializes Talk's `talkHashStore` with the hash restored with `appData`
- Subscribe Talk's `setInitialNextcloudTalkHash` mutation to save initial hash
- Subscribe Talk's `markNextcloudTalkHashDirty` mutation to handle invalide hash and relaunch the app

Possible scenarios:

#### Scenario 1 - appData invalidates between launches

1. App restores the appData with the old hash
2. The old hash is set to the Talk's store
3. On Talk app init (conversations loading) it receives new hash
4. Hash is dirty - save it to the appData
5. App is relaunching
6. On relaunch during the welcome screen new capabilities and userMetadata are fetched and persisted

#### Scenario 2 - appData invalidates after the launch

The same

#### Scenario 3 - old appData before this update

The same. Having now `talkHashDirty` flag considers dirty.

## Drawbacks and alternative solutions

1. Show a message instead of the hard relaunch
   - On hash invalidate Talk web shows a toast asking to reload and makes a web page barely working
   - It requires patching Talk web to change the toast behavior
   - But in Talk Web, we cannot call, save drafts or join chats on invalidate anyway
   - Drawbacks:
     - Possibility of saving a draft of the current chats (other drafts are lost anyway)
     - Hard relaunch doesn't look good in UX
   - I prefer the current solution on desktop and think, the game is not worth the candle
2. Always re-fetch new appData on launch
   - Currently, if appData is invalid, it would be known only after launch and fetching conversations
   - It makes the launch fast, but relaunches app on invalidates
   - Alternative - make launch always slower, but without relaunch on invalidate between launches
   - I prefer the current solution because a typical launch doesn't require re-fetch

### CODE REVIEW

File: src/app/appData.service.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { getCurrentUserData, getCapabilities } from '../shared/ocs.service.js'
+
+/**
+ * Re-fetch capabilities and userMetadata and update appData
+ *
+ * @param {AppData} appData
```
Comment (nickvergessen) [inline src/app/appData.service.js:27]: Some eslint mimi here and there

File: src/app/appData.service.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	appData.version.talk = talkCapabilities.version
+	if (persist) {
+		appData.persist()
+	}
+}
+
+/**
+ * If talk hash is dirty, re-fetch capabilities and userMetadata and update appData
+ *
+ * @param {AppData} appData
```
Comment (nickvergessen) [inline src/app/appData.service.js:52]: Same here

File: src/authentication/renderer/AuthenticationApp.vue [hunk 3]
```diff
@@ -77,6 +77,7 @@ import NcTextField from '@nextcloud/vue/dist/Components/NcTextField.js'
 import { getCapabilities, getCurrentUserData } from '../../shared/ocs.service.js'
```
Comment (nickvergessen) [inline src/authentication/renderer/AuthenticationApp.vue:77]: 'getCurrentUserData' is defined but never used

```suggestion
import { getCapabilities } from '../../shared/ocs.service.js'
```

**Review by nickvergessen — COMMENTED** [review]
Can't read it from the code really, so just commenting here:
In case it is not happening already, we should also refetch the data if the desktop client version changed. So that when people update the client they always get newest data.

**Comment by ShGKme** [comment]
> In case it is not happening already, we should also refetch the data if the desktop client version changed. So that when people update the client they always get newest data.

Do you mean, on any update, even if a talk hash was not changed?

**Comment by ShGKme** [comment]
> yes, if the talk client version changes, refetch the data. Maybe at some point we don't expose a feature to specific client versions or something alike.

Done in the last commit

**Review by nickvergessen — APPROVED** [review]
Didn't test, but looks good

**Comment by ShGKme** [comment]
Small update, also save new desktop client version in the appData on re-fetch

**Pull request merged**